File: D/2095-delete-the-middle-node-of-a-linked-list/main.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteMiddle(head *ListNode) *ListNode {
	p := head
	cnt := 0
	for p != nil {
		cnt++
		p = p.Next
	}
	steps := cnt / 2
	NH := ListNode{Val: 0, Next: head}
	pp := &NH
	p = head
	for steps > 0 {
		steps--
		pp = p
		p = p.Next
	}
	pp.Next = p.Next
	return NH.Next
}

func main() {
	fmt.Println()
}

File: D/2095-delete-the-middle-node-of-a-linked-list/main.rs
impl Solution {
  pub fn delete_middle(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut h = &head;
    let mut cnt: i32 = 0;
    while let Some(ln) = h {
      cnt += 1;
      h = &ln.next;
    }

    cnt /= 2;
    let mut nh = Box::new(ListNode { val: 0, next: head });
    let mut pp = &mut nh;
    while cnt > 0 {
      pp = pp.next.as_mut()?;
      cnt -= 1;
    }
    pp.next = pp.next.as_mut()?.next.take();

    nh.next
  }
}

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
